File: handlers/game/events.go
package game

import (
	"github.com/cannonball10/foundation/schemas/replicant"
)

// ProgressReason names why the engine advanced a phase. It is attached
// to transition events so clients can render "cutoff", "forced", etc.
type ProgressReason string

const (
	ReasonAction   ProgressReason = "action"    // normal player/host action
	ReasonTimeout  ProgressReason = "timeout"   // deadline expired
	ReasonForced   ProgressReason = "forced"    // host forced progress
	ReasonAllVoted ProgressReason = "all_voted" // every alive player voted
)

// --- typed event payloads --------------------------------------------------
//
// Each struct below corresponds to a replicant.EventType. They are
// attached to the Envelope.Payload field so clients get typed data
// without needing to re-interpret the map[string]any on GameEvent.Data.

// GameCreatedPayload announces a new lobby.
type GameCreatedPayload struct {
	JoinCode   string `json:"joinCode"`
	HostUserID string `json:"hostUserId"`
}

// PlayerJoinedPayload announces a player entering the lobby.
type PlayerJoinedPayload struct {
	PlayerID    string `json:"playerId"`
	DisplayName string `json:"displayName"`
	Seat        int    `json:"seat"`
}

// GameStartedPayload announces that the game has begun. Seating order
// and the initial president are fixed by this event.
type GameStartedPayload struct {
	PlayerCount          int `json:"playerCount"`
	InitialPresidentSeat int `json:"initialPresidentSeat"`
	Round                int `json:"round"`
}

// RoleAssignedPayload is whispered individually to each player so they
// learn their own role (and, for fascists/Hitler in small games, their
// teammates).
type RoleAssignedPayload struct {
	Role      replicant.Role  `json:"role"`
	Party     replicant.Party `json:"party"`
	Teammates []TeammateInfo     `json:"teammates,omitempty"`
}

// TeammateInfo is the private view a fascist has of the cabal at game
// start (or Hitler, in small games).
type TeammateInfo struct {
	PlayerID    string            `json:"playerId"`
	DisplayName string            `json:"displayName"`
	Role        replicant.Role `json:"role"`
}

// ChancellorNominatedPayload accompanies a chancellor nomination. Round
// is included so clients can display the round counter without fetching
// the full game snapshot; it changes exactly when a new round begins.
type ChancellorNominatedPayload struct {
	PresidentPlayerID  string `json:"presidentPlayerId"`
	ChancellorPlayerID string `json:"chancellorPlayerId"`
	GovernmentID       string `json:"governmentId"`
	Deadline           string `json:"deadline,omitempty"` // RFC3339
	Round              int    `json:"round"`
}

// VoteCastPayload is broadcast when a player votes. The choice is NOT
// included; only the fact that the player voted (so everyone can see
// the vote tally fill up without spoilers).
type VoteCastPayload struct {
	GovernmentID string `json:"governmentId"`
	PlayerID     string `json:"playerId"`
}

// ElectionResultPayload reports the final vote tally.
type ElectionResultPayload struct {
	GovernmentID string                             `json:"governmentId"`
	Passed       bool                               `json:"passed"`
	JaVotes      int                                `json:"jaVotes"`
	NeinVotes    int                                `json:"neinVotes"`
	Votes        map[string]replicant.VoteChoice `json:"votes"` // keyed by playerId
	Reason       ProgressReason                     `json:"reason"`
}

// PoliciesDrawnPayload is whispered to the president after a successful
// election so they see the three policies they must choose from.
type PoliciesDrawnPayload struct {
	GovernmentID string                    `json:"governmentId"`
	Policies     []replicant.PolicyType `json:"policies"`
}

// PresidentDiscardedPayload is broadcast (counts only) and whispered to
// the chancellor (with the two remaining policies).
type PresidentDiscardedPayload struct {
	GovernmentID string                    `json:"governmentId"`
	Options      []replicant.PolicyType `json:"options,omitempty"` // chancellor-only
}

// ChancellorEnactedPayload is broadcast when a policy is placed on the board.
type ChancellorEnactedPayload struct {
	GovernmentID           string                  `json:"governmentId"`
	Policy                 replicant.PolicyType `json:"policy"`
	HumanPoliciesEnacted int                     `json:"humanPoliciesEnacted"`
	AIPoliciesEnacted int                     `json:"aiPoliciesEnacted"`
}

// VetoProposedPayload announces the chancellor requested a veto.
type VetoProposedPayload struct {
	GovernmentID string `json:"governmentId"`
}

// VetoResolvedPayload announces the president's response.
type VetoResolvedPayload struct {
	GovernmentID string `json:"governmentId"`
	Accepted     bool   `json:"accepted"`
}

// ExecutiveActionPayload announces the use of a presidential power. For
// policy peek the actual policies are whispered separately to the
// president; for investigate, the target's party is whispered.
type ExecutiveActionPayload struct {
	ActionID          string                           `json:"actionId"`
	Type              replicant.ExecutiveActionType `json:"type"`
	PresidentPlayerID string                           `json:"presidentPlayerId"`
	TargetPlayerID    string                           `json:"targetPlayerId,omitempty"`
}

// InvestigateResultPayload is whispered to the president after an
// Investigate Loyalty action.
type InvestigateResultPayload struct {
	ActionID       string             `json:"actionId"`
	TargetPlayerID string             `json:"targetPlayerId"`
	Party          replicant.Party `json:"party"`
}

// PolicyPeekPayload is whispered to the president after a Policy Peek.
type PolicyPeekPayload struct {
	ActionID string                    `json:"actionId"`
	Policies []replicant.PolicyType `json:"policies"` // top 3, in order
}

// ElectionTrackerPayload accompanies a tracker advance after a failed
// election or veto.
type ElectionTrackerPayload struct {
	Tracker int `json:"tracker"`
}

// TopDeckPayload announces a policy enacted by election-tracker advance.
type TopDeckPayload struct {
	Policy                 replicant.PolicyType `json:"policy"`
	HumanPoliciesEnacted int                     `json:"humanPoliciesEnacted"`
	AIPoliciesEnacted int                     `json:"aiPoliciesEnacted"`
}

// DeckReshuffledPayload announces the draw+discard piles were combined
// and reshuffled (happens whenever the draw pile has < 3 cards).
type DeckReshuffledPayload struct {
	RemainingInDraw int `json:"remainingInDraw"`
}

// PlayerExecutedPayload announces a player was executed by the
// Execution power.
type PlayerExecutedPayload struct {
	PlayerID       string `json:"playerId"`
	WasRogue      bool   `json:"wasRogue"`
	ExecutedBySeat int    `json:"executedBySeat"`
}

// GameEndedPayload announces the final result.
type GameEndedPayload struct {
	Winner       replicant.Party        `json:"winner"`
	WinCondition replicant.WinCondition `json:"winCondition"`
}

// NarratorSpeakPayload carries a synthesised narrator utterance. The
// Committee "speaks" over a game-wide broadcast envelope; host
// devices play AudioURL (a data-URL embedding the MP3) and render
// Script as typewriter text. Mobile clients can optionally render the
// transcript without playing audio.
type NarratorSpeakPayload struct {
	CueID    string `json:"cueId"`
	Cue      string `json:"cue"`
	Script   string `json:"script"`
	AudioURL string `json:"audioUrl"`
	AudioID  string `json:"audioId"`
	Format   string `json:"format"`
	TookMS   int64  `json:"tookMs"`
}

// CablePhaseOpenedPayload announces the start of a Cable Phase. The
// deadline is the phase's absolute cutoff in RFC3339 — clients render
// a countdown from (deadline - now). Chat send controls should become
// live on receipt and lock again on cable_phase_closed.
type CablePhaseOpenedPayload struct {
	GovernmentID string `json:"governmentId"`
	Deadline     string `json:"deadline"`
}

// CablePhaseClosedPayload announces the end of a Cable Phase. Reason
// mirrors ProgressReason (timeout / forced / all_voted-equivalent).
// LeakedMessageID references the cable (if any) the Committee chose
// to broadcast; empty when the Committee went silent or no cables
// were submitted.
type CablePhaseClosedPayload struct {
	GovernmentID     string         `json:"governmentId"`
	Reason           ProgressReason `json:"reason"`
	LeakedMessageID  string         `json:"leakedMessageId,omitempty"`
	TotalSubmissions int            `json:"totalSubmissions"`
}

// CableLeakedPayload carries the full body of the cable the narrator
// just broadcast — including a flag indicating whether the Committee
// chose silence instead of a leak. When Silenced is true, Body and
// Author are empty; the narrator_speak envelope fired alongside
// announces the silence verbally.
type CableLeakedPayload struct {
	GovernmentID    string  `json:"governmentId"`
	Silenced        bool    `json:"silenced"`
	MessageID       string  `json:"messageId,omitempty"`
	AuthorPlayerID  string  `json:"authorPlayerId,omitempty"`
	Author          string  `json:"author,omitempty"`
	Body            string  `json:"body,omitempty"`
	SubversionScore float64 `json:"subversionScore,omitempty"`
}

// SingularityCableFeedPayload is whispered only to the Singularity at
// cable-phase-close: every cable submitted this round, anonymised.
// The kingmaker role's intel edge — full chatter, no attribution.
type SingularityCableFeedPayload struct {
	GovernmentID string                    `json:"governmentId"`
	Cables       []AnonymousCableFeedEntry `json:"cables"`
}

// AnonymousCableFeedEntry is one row of the Singularity feed: message
// id + body + (optional) subversion score, with no author.
type AnonymousCableFeedEntry struct {
	MessageID       string  `json:"messageId"`
	Body            string  `json:"body"`
	SubversionScore float64 `json:"subversionScore,omitempty"`
}

// PhaseChangedPayload is emitted on every phase transition so clients
// can drive their UI from a single event stream. It pairs nicely with
// more specific events (e.g. ChancellorNominatedPayload) but stands on
// its own when the engine is advanced by a timer or host action.
//
// PresidentSeat is the CURRENT president at the time of the transition.
// Clients use it to recover the seat after rotation (normal round
// rollover) and after a Special Election, neither of which carry a
// dedicated "president changed" event.
type PhaseChangedPayload struct {
	From          replicant.GamePhase `json:"from"`
	To            replicant.GamePhase `json:"to"`
	Reason        ProgressReason         `json:"reason"`
	Deadline      string                 `json:"deadline,omitempty"` // RFC3339
	PresidentSeat int                    `json:"presidentSeat"`
}

File: packages/schema/src/events.ts
// Envelope wire format — mirrors handlers/game/events.go payloads and
// handlers/game/emitter.go envelope shape. Every SSE frame from the
// server parses into one of these Envelope variants.

import { z } from "zod";
import {
  AudienceScope,
  ChatChannel,
  EventType,
  ExecutiveActionType,
  GamePhase,
  Party,
  PolicyType,
  ProgressReason,
  Role,
  VoteChoice,
  WinCondition,
} from "./enums";

// ─── audience ────────────────────────────────────────────────────

export const Audience = z.object({
  scope: AudienceScope,
  playerId: z.string().optional(),
});
export type Audience = z.infer<typeof Audience>;

// ─── individual payloads ─────────────────────────────────────────

export const GameCreatedPayload = z.object({
  joinCode: z.string(),
  hostUserId: z.string(),
});
export type GameCreatedPayload = z.infer<typeof GameCreatedPayload>;

export const PlayerJoinedPayload = z.object({
  playerId: z.string(),
  displayName: z.string(),
  seat: z.number(),
});
export type PlayerJoinedPayload = z.infer<typeof PlayerJoinedPayload>;

export const GameStartedPayload = z.object({
  playerCount: z.number(),
  initialPresidentSeat: z.number(),
  round: z.number(),
});
export type GameStartedPayload = z.infer<typeof GameStartedPayload>;

export const TeammateInfo = z.object({
  playerId: z.string(),
  displayName: z.string(),
  role: Role,
});
export type TeammateInfo = z.infer<typeof TeammateInfo>;

export const RoleAssignedPayload = z.object({
  role: Role,
  party: Party,
  teammates: z.array(TeammateInfo).optional(),
});
export type RoleAssignedPayload = z.infer<typeof RoleAssignedPayload>;

export const ChancellorNominatedPayload = z.object({
  presidentPlayerId: z.string(),
  chancellorPlayerId: z.string(),
  governmentId: z.string(),
  deadline: z.string().optional(),
  round: z.number(),
});
export type ChancellorNominatedPayload = z.infer<typeof ChancellorNominatedPayload>;

export const VoteCastPayload = z.object({
  governmentId: z.string(),
  playerId: z.string(),
});
export type VoteCastPayload = z.infer<typeof VoteCastPayload>;

export const ElectionResultPayload = z.object({
  governmentId: z.string(),
  passed: z.boolean(),
  jaVotes: z.number(),
  neinVotes: z.number(),
  votes: z.record(VoteChoice),
  reason: ProgressReason,
});
export type ElectionResultPayload = z.infer<typeof ElectionResultPayload>;

export const PoliciesDrawnPayload = z.object({
  governmentId: z.string(),
  policies: z.array(PolicyType),
});
export type PoliciesDrawnPayload = z.infer<typeof PoliciesDrawnPayload>;

export const PresidentDiscardedPayload = z.object({
  governmentId: z.string(),
  options: z.array(PolicyType).optional(),
});
export type PresidentDiscardedPayload = z.infer<typeof PresidentDiscardedPayload>;

export const ChancellorEnactedPayload = z.object({
  governmentId: z.string(),
  policy: PolicyType,
  humanPoliciesEnacted: z.number(),
  aiPoliciesEnacted: z.number(),
});
export type ChancellorEnactedPayload = z.infer<typeof ChancellorEnactedPayload>;

export const VetoProposedPayload = z.object({ governmentId: z.string() });
export type VetoProposedPayload = z.infer<typeof VetoProposedPayload>;

export const VetoResolvedPayload = z.object({
  governmentId: z.string(),
  accepted: z.boolean(),
});
export type VetoResolvedPayload = z.infer<typeof VetoResolvedPayload>;

export const ExecutiveActionPayload = z.object({
  actionId: z.string(),
  type: ExecutiveActionType,
  presidentPlayerId: z.string(),
  targetPlayerId: z.string().optional(),
});
export type ExecutiveActionPayload = z.infer<typeof ExecutiveActionPayload>;

export const InvestigateResultPayload = z.object({
  actionId: z.string(),
  targetPlayerId: z.string(),
  party: Party,
});
export type InvestigateResultPayload = z.infer<typeof InvestigateResultPayload>;

export const PolicyPeekPayload = z.object({
  actionId: z.string(),
  policies: z.array(PolicyType),
});
export type PolicyPeekPayload = z.infer<typeof PolicyPeekPayload>;

export const ElectionTrackerPayload = z.object({ tracker: z.number() });
export type ElectionTrackerPayload = z.infer<typeof ElectionTrackerPayload>;

export const TopDeckPayload = z.object({
  policy: PolicyType,
  humanPoliciesEnacted: z.number(),
  aiPoliciesEnacted: z.number(),
});
export type TopDeckPayload = z.infer<typeof TopDeckPayload>;

export const DeckReshuffledPayload = z.object({ remainingInDraw: z.number() });
export type DeckReshuffledPayload = z.infer<typeof DeckReshuffledPayload>;

export const PlayerExecutedPayload = z.object({
  playerId: z.string(),
  wasRogue: z.boolean(),
  executedBySeat: z.number(),
});
export type PlayerExecutedPayload = z.infer<typeof PlayerExecutedPayload>;

export const GameEndedPayload = z.object({
  winner: Party,
  winCondition: WinCondition,
});
export type GameEndedPayload = z.infer<typeof GameEndedPayload>;

export const ChatMessagePayload = z.object({
  messageId: z.string(),
  channel: ChatChannel,
  authorPlayerId: z.string(),
  authorDisplayName: z.string(),
  body: z.string(),
  sentAt: z.string(),
  governmentId: z.string().optional(),
  ack: z.boolean().optional(),
});
export type ChatMessagePayload = z.infer<typeof ChatMessagePayload>;

// CablePhaseOpenedPayload — server's signal to unlock cable compose
// on every eligible device. Deadline is RFC3339 and drives the
// mobile countdown.
export const CablePhaseOpenedPayload = z.object({
  governmentId: z.string(),
  deadline: z.string(),
});
export type CablePhaseOpenedPayload = z.infer<typeof CablePhaseOpenedPayload>;

// CablePhaseClosedPayload — server's signal to lock cable compose.
// `leakedMessageId` is populated when the narrator leaks a cable,
// and matches the `messageId` in the follow-up cable_leaked envelope.
export const CablePhaseClosedPayload = z.object({
  governmentId: z.string(),
  reason: ProgressReason,
  leakedMessageId: z.string().optional(),
  totalSubmissions: z.number(),
});
export type CablePhaseClosedPayload = z.infer<typeof CablePhaseClosedPayload>;

// CableLeakedPayload — the narrator's phase-end broadcast. When
// `silenced` is true the body/author are empty and the paired
// narrator_speak envelope says the Committee reviewed the traffic
// and found nothing worth flagging.
export const CableLeakedPayload = z.object({
  governmentId: z.string(),
  silenced: z.boolean(),
  messageId: z.string().optional(),
  authorPlayerId: z.string().optional(),
  author: z.string().optional(),
  body: z.string().optional(),
  subversionScore: z.number().optional(),
});
export type CableLeakedPayload = z.infer<typeof CableLeakedPayload>;

/**
 * NarratorSpeakPayload — the Committee's synthesised utterance.
 * `audioUrl` is a `data:audio/mpeg;base64,…` URL; host clients play
 * it directly (and may GET /api/v1/narrator/audio/:audioId as a
 * fallback if they need to re-fetch). Mobile clients may render only
 * `script` if they prefer silent transcripts.
 */
export const NarratorSpeakPayload = z.object({
  cueId: z.string(),
  cue: z.string(),
  script: z.string(),
  audioUrl: z.string(),
  audioId: z.string(),
  format: z.string().optional(),
  tookMs: z.number().optional(),
});
export type NarratorSpeakPayload = z.infer<typeof NarratorSpeakPayload>;

/**
 * Phase-change events reuse the election_result envelope type on the
 * wire (see handlers/game/engine.go:setPhase). Carries the current
 * presidentSeat so clients can pick up rotation.
 */
export const PhaseChangedPayload = z.object({
  from: GamePhase,
  to: GamePhase,
  reason: ProgressReason,
  deadline: z.string().optional(),
  presidentSeat: z.number(),
});
export type PhaseChangedPayload = z.infer<typeof PhaseChangedPayload>;

// ─── envelope ────────────────────────────────────────────────────

/**
 * The GameEvent record carried inside every Envelope.event. Free-form
 * `data` is a lower-fidelity copy of the typed payload — we prefer the
 * payload field in reducers and fall back on `data` only when a newer
 * server adds a payload type we haven't mapped yet.
 */
export const GameEvent = z.object({
  eventId: z.string(),
  gameId: z.string(),
  type: EventType,
  actorPlayerId: z.string().optional(),
  targetPlayerId: z.string().optional(),
  data: z.record(z.unknown()).optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
});
export type GameEvent = z.infer<typeof GameEvent>;

export const Envelope = z.object({
  gameId: z.string(),
  event: GameEvent,
  audience: Audience,
  payload: z.unknown().optional(),
});
export type Envelope = z.infer<typeof Envelope>;
